### PR TITLE
Refactoring, more tests and more coverage

### PR DESCRIPTION
**Pull request by vihu**
* Add comparison functions to poly and bipoly
* Code refactor
* Fix more typos
* Add public_key_share_test, comparison tests
* Update README a bit

### CODE REVIEW

File: src/dkg_bipolynomial.erl [hunk 1]
```diff
@@ -27,9 +36,23 @@ add(PolyA, PolyB) ->
...
+cmp(PolyA, PolyB) ->
+    %% checks f(x, y) - g(x, y) = 0
+    %% subtracting a polynomial from itself should yield all 0s
+    %% and all the coefficients should match
+    lists:all(fun(X) ->
+                      X
+              end,
+              [ erlang_pbc:element_cmp(lookup([I, J], PolyA), lookup([I, J], PolyB)) || I <- lists:seq(1, degree(PolyA)), J <- lists:seq(1, degree(PolyB))])
+    andalso
+    is_zero(dkg_bipolynomial:sub(PolyA, PolyB)).
```
Comment (Vagabond) [inline src/dkg_bipolynomial.erl]: Why do both a compare and subtract? You should also check they both have the same degree or the list comprehension will fail.

File: src/dkg_commitmentmatrix.erl [hunk 2]
```diff
@@ -82,7 +84,6 @@ public_key_share(Matrix, NodeID) ->
                         R = erlang_pbc:element_pow(Acc, M),
                         Row = lists:foldl(fun(J, Acc2) ->
                                                   erlang_pbc:element_mul(erlang_pbc:element_pow(Acc2, I), lookup([II, J], Matrix))
-                                          end, G1, lists:seq(tuple_size(Matrix))),
+                                          end, G1, lists:seq(1, tuple_size(Matrix))),
                         erlang_pbc:element_mul(R, Row)
-                end, G1, lists:seq(tuple_size(Matrix))).
-
+                end, G1, lists:seq(1, tuple_size(Matrix))).
```
Comment (Vagabond) [inline src/dkg_commitmentmatrix.erl]: I think both these folds need to go in reverse as elsewhere?

File: src/dkg_polynomial.erl [hunk 3]
```diff
@@ -35,6 +63,14 @@ mul(PolyA, PolyB) ->
                         add(Acc, Temp)
                 end, [], PolyB).
 
+cmp(PolyA, PolyB) ->
+    %% check f(x) - g(x) = 0,
+    %% check that each element is the same the original
+    %% and the degree of polyA = polyB
+    dkg_polynomial:is_zero(dkg_polynomial:sub(PolyA, PolyB)) andalso
+    is_equal(PolyA, PolyB) andalso
+    degree(PolyA) == degree(PolyB).
```
Comment (Vagabond) [inline src/dkg_polynomial.erl]: Zip in the previous step will fail if the 2 lengths are not the same length.

Comment (Vagabond) [reply src/dkg_polynomial.erl]: I'd do the degree check first, and remove the zero check?

File: test/dkg_commitmentmatrix_SUITE.erl [hunk 4]
```diff
@@ -16,6 +24,12 @@ init_per_testcase(_, Config) ->
 end_per_testcase(_, Config) ->
     Config.
 
+new_matrix_test(Config) ->
+    Pairing = proplists:get_value(pairing, Config),
+    BiPoly = dkg_bipolynomial:generate(Pairing, 4),
+    io:format("BiPoly: ~p~n", [dkg_bipolynomial:print(BiPoly)]),
```
Comment (Vagabond) [inline test/dkg_commitmentmatrix_SUITE.erl]: What is this testing?

File: src/dkg_polynomial.erl [hunk 5]
```diff
@@ -20,6 +31,23 @@ generate(Pairing, T, Index, Term) ->
     NewHead = erlang_pbc:element_add(erlang_pbc:element_sub(Head, Val), Term),
     [NewHead | Tail].
 
+degree(Poly) ->
```
Comment (Vagabond) [inline src/dkg_polynomial.erl]: Still uncomfortable with this change. Shouldn't we stick to the mathematical definition?

File: test/dkg_commitmentmatrix_SUITE.erl (hunk 4 above)
Comment (vihu) [reply test/dkg_commitmentmatrix_SUITE.erl]: Oh, this creeped in, I'll remove this test, my intention was to test the identity matrix, but can do it in another branch.

File: src/dkg_polynomial.erl (hunk 3 above)
Comment (vihu) [reply src/dkg_polynomial.erl]: Agreed on the degree check first. So the _formal_ definition of whether f(x) = g(x) states that if f(x) - g(x) = h(x), then h(x) = 0 for all values of x. Hence the zero checks for both poly and bipoly.

File: src/dkg_bipolynomial.erl (hunk 1 above)
Comment (vihu) [reply src/dkg_bipolynomial.erl]: Agreed on checking the degree first, the need for zero test is the same as for poly test, basically f(x, y) = g(x, y) if f(x, y) - g(x, y) = h(x, y); then h(x, y) is 0 for all values of x and y. Atlteast theoretically.

File: src/dkg_polynomial.erl (hunk 3 above)
Comment (Vagabond) [reply src/dkg_polynomial.erl]: Why is that better than just straight element comparison, though? I can see that being useful in a mathematical way, but we don't subtract a number from itself to check if it's the same in programming?

Comment (vihu) [reply src/dkg_polynomial.erl]: Fair enough, I was just being pedantic with the mathematical definition.

**Pull request merged**